Utils/src/GenParticlesProducer.cc

```diff
@@ -302,42 +304,66 @@ void GenParticlesProducer::storeMinimal(const edm::Handle< edm::View<reco::GenPa
                                         std::unique_ptr<std::vector<int>>& ParentId_vec) const {
     std::unordered_set<const reco::Candidate *> stored_particles_ref;
     std::vector<const reco::Candidate *> stored_particles_list, parents_list;
-    for(const auto& iPart : *genPartCands) {
+    for(const auto& particle : *genPartCands) {
         // Skip particles which are already in the list of stored particles
         // This is to protect against parts of a chain which may be saved, but here we are looking at the first copy
-        if(stored_particles_ref.find(&iPart)!=stored_particles_ref.end()) continue;
+        if(stored_particles_ref.find(&particle)!=stored_particles_ref.end()) continue;
+
+        // Store the initial partons
+        auto nmothers = particle.numberOfMothers();
+        if(nmothers == 1) {
+           auto mother = particle.mother(0);
+           if(mother->pdgId() == proton &&
+              std::binary_search(proton_status_codes.begin(), proton_status_codes.end(), mother->status()) &&
+              (mother == &(*genPartCands->begin()) || mother == &(*(genPartCands->begin()+1))) &&
```

Comment:
could keep the comparison pointers outside the loop:
```cpp
std::array<const reco::Candidate*, 2> possible_mothers{&(*genPartCands->begin()), &(*(genPartCands->begin()+1)))};
```
and then compare to `possible_mothers[0]` and `possible_mothers[1]` inside the loop (avoid repeated creation of iterators, dereferencing, etc.)

Reply:
Done